qutebrowser/misc/sessions.py

```diff
@@ -428,6 +440,10 @@ def load(self, name, temp=False):
                     tab_to_focus = i
                 if new_tab.data.pinned:
                     tabbed_browser.set_tab_pinned(new_tab, new_tab.data.pinned)
+                if config.val.tabs.mode_on_change == 'restore':
+                    modeman.enter(window.win_id,
+                                  new_tab.data.input_mode,
+                                  'restore input mode for tab')
```

Comment:
Why are you doing this for every tab? Also, shouldn't the focusing take care of this already anyways?

Reply:
Yeah, I guess it still needs investigation as I cannot make it work otherwise. I was under the same impression about focusing, but somehow this is not working just by focusing as I would expect.

I think when the tabs are restored, they sort of become active (and focused) for a moment, so the input_mode gets reset somehow, so I need to force it. But I'll dig deeper.